parser/src/main/scala/jawn/Slice.scala

```diff
@@ -0,0 +1,55 @@
+package jawn
+
+/**
+ * Character sequence representing a lazily-calculated substring.
+ *
+ * This class has three constructors:
+ *
+ *  - Slice(s) wraps a string, ensuring that future operations
+ *    (e.g. subSequence) will construct slices instead of strings.
+ *
+ *  - Slice(s, start, limit) is the default, and ensures that:
+ *
+ *     1. start >= 0
+ *     2. limit >= start
+ *     3. limit <= s.length
+ *
+ *  - Slice.unsafe(s, start, limit) is for situations where the above
+ *    bounds-checking has already occurred. Only use this if you are
+ *    absolutely sure your arguments satisfy the above invariants.
+ *
+ * Slice's subSequence returns another slice. This means that when
+ * wrapping a very large string, garbage collection on the underlying
+ * string will not occur until all slices are freed.
+ */
+final class Slice private[jawn] (s: String, start: Int, limit: Int) extends CharSequence {
```

Comment:
Also I'm not sure it's necessary but I'm now paranoid about `Serializable`, and I could imagine that if circe ended up using these in its JSON string representation we'd want them to be `Serializable`.